File: Components/DevExpressVCL/Demos/ExpressPrinting System/Delphi/PivotGridRL/PivotGridRLMain.pas
unit PivotGridRLMain;

interface

{$I cxVer.inc}

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, DemoBasicMain, dxPSGlbl, dxPSUtl, dxPSEngn, dxPrnPg, dxBkgnd,
  dxWrap, dxPrnDev, dxPSCompsProvider, dxPSFillPatterns, dxPSEdgePatterns,
  dxPSPDFExportCore, dxPSPDFExport, cxDrawTextUtils, dxPSPrVwStd,
  dxPScxEditorProducers, dxPScxExtEditorProducers,
  dxPScxPageControlProducer, dxPSCore, ActnList, ImgList, Menus, ComCtrls,
  ToolWin, StdCtrls, cxGraphics, cxControls, cxLookAndFeels,
  cxLookAndFeelPainters, cxClasses, cxCustomData, cxStyles, cxEdit,
  cxCustomPivotGrid, cxPivotGrid, dxPScxCommon, dxPScxPivotGridLnk, dxCore;

type
  TPivotGridRLMainForm = class(TDemoBasicMainForm)
    PaymentTypeImages: TImageList;
    PivotGrid: TcxPivotGrid;
    pgfPurchaseQuarter: TcxPivotGridField;
    pgfPurchaseMonth: TcxPivotGridField;
    pgfPaymentType: TcxPivotGridField;
    pgfQuantity: TcxPivotGridField;
    pgfCarName: TcxPivotGridField;
    pgfUnitPrice: TcxPivotGridField;
    pgfCompanyName: TcxPivotGridField;
    pgfPaymentAmount: TcxPivotGridField;
    dxComponentPrinterLink1: TcxPivotGridReportLink;
    procedure FormCreate(Sender: TObject);
    procedure pgfPaymentTypeGetGroupImageIndex(Sender: TcxPivotGridField;
      const AItem: TcxPivotGridViewDataItem; var AImageIndex: Integer;
      var AImageAlignHorz: TAlignment;
      var AImageAlignVert: TcxAlignmentVert);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  PivotGridRLMainForm: TPivotGridRLMainForm;

implementation

{$R *.dfm}

function EnumResModules(Instance: TdxNativeUInt; Data: Pointer): Boolean;
begin
  Result := True;
  if FindResource(Instance, 'PIVOTPREVIEWDATA','PIVOTDATA') <> 0 then
  begin
    PInteger(Data)^ := Instance;
    Result := False;
  end;
end;

procedure TPivotGridRLMainForm.FormCreate(Sender: TObject);
var
  AStream: TStream;
  AInstance: TdxNativeUInt;
begin
  inherited;
  AInstance := 0;
  EnumResourceModules(EnumResModules, @AInstance);
  AStream := TResourceStream.Create(AInstance, 'PIVOTPREVIEWDATA', 'PIVOTDATA');
  try
    AStream.Position := 0;
    PivotGrid.DataController.LoadFromStream(AStream);
  finally
    AStream.Free;
  end;
end;

procedure TPivotGridRLMainForm.pgfPaymentTypeGetGroupImageIndex(
  Sender: TcxPivotGridField; const AItem: TcxPivotGridViewDataItem;
  var AImageIndex: Integer; var AImageAlignHorz: TAlignment;
  var AImageAlignVert: TcxAlignmentVert);
var
  ACard: string;
begin
  inherited;
  ACard := VarToStr(AItem.Value);
  if SameText(ACard, 'AmEx') then
    AImageIndex := 0
  else
    if SameText(ACard, 'Cash') then
      AImageIndex := 1
    else
      if SameText(ACard, 'Master') then
        AImageIndex := 2
      else
        if SameText(ACard, 'Visa') then
          AImageIndex := 3;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressPrinting System/CBuilder/PivotGridRL/PivotGridRLMain.cpp
// ---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop
#include "PivotGridRLMain.h"
// ---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxControls"
#pragma link "cxCustomPivotGrid"
#pragma link "cxPivotGrid"
#pragma link "dxPSCore"
#pragma link "dxPScxCommon"
#pragma link "dxPScxPivotGridLnk"
#pragma link "cxClasses"
#pragma link "cxCustomData"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxStyles"
#pragma link "cxDrawTextUtils"
#pragma link "DemoBasicMain"
#pragma link "dxBkgnd"
#pragma link "dxPrnDev"
#pragma link "dxPrnPg"
#pragma link "dxPSCompsProvider"
#pragma link "dxPScxEditorProducers"
#pragma link "dxPScxExtEditorProducers"
#pragma link "dxPScxPageControlProducer"
#pragma link "dxPSEdgePatterns"
#pragma link "dxPSEngn"
#pragma link "dxPSFillPatterns"
#pragma link "dxPSGlbl"
#pragma link "dxPSPDFExport"
#pragma link "dxPSPDFExportCore"
#pragma link "dxPSPrVwStd"
#pragma link "dxPSUtl"
#pragma link "dxWrap"
#pragma resource "*.dfm"
TPivotGridRLMainForm *PivotGridRLMainForm;

// ---------------------------------------------------------------------------
__fastcall TPivotGridRLMainForm::TPivotGridRLMainForm(TComponent* Owner)
: TDemoBasicMainForm(Owner) {
}

// ---------------------------------------------------------------------------
void __fastcall TPivotGridRLMainForm::FormCreate(TObject *Sender) {

	TDemoBasicMainForm::FormCreate(Sender);
#if defined(_DELPHI_STRING_UNICODE)
	PChar S = PChar(WideString("PIVOTDATA"));
#else
	PChar S = "PIVOTDATA";
#endif
	int AInstance;
	TStream * AStream = new TFileStream("PIVOTPREVIEWDATA.DAT", fmOpenRead);
	try {
		AStream->Position = 0;
		PivotGrid->DataController->LoadFromStream(AStream);
	}
	__finally {
		delete AStream;
	}
}
// ---------------------------------------------------------------------------

void __fastcall TPivotGridRLMainForm::pgfPaymentTypeGetGroupImageIndex
(TcxPivotGridField *Sender, const TcxPivotGridViewDataItem *AItem,
	int &AImageIndex, TAlignment &AImageAlignHorz, TcxAlignmentVert &AImageAlignVert) {
	AnsiString Card = ((TcxPivotGridViewDataItem*)AItem)->Value;
	if (SameText(Card, "AmEx"))
		AImageIndex = 0;
	else if (SameText(Card, "Cash"))
		AImageIndex = 1;
	else if (SameText(Card, "Master"))
		AImageIndex = 2;
	else if (SameText(Card, "Visa"))
		AImageIndex = 3;
}
// ---------------------------------------------------------------------------
